Buildings/Fluid/HeatExchangers/BaseClasses/WetCoilUARated.mo

```diff
@@ -0,0 +1,189 @@
+
+  Modelica.SIunits.AbsolutePressure pSatTWatIn=
+    Buildings.Utilities.Psychrometrics.Functions.saturationPressure(TWatIn)
+    "Saturation pressure of water at the water inlet temperature";
+  Modelica.SIunits.MassFraction X_wSatTWatIn=
+    Buildings.Utilities.Psychrometrics.Functions.X_pSatpphi(
+      pSat=pSatTWatIn, p=MediumA.p_default, phi=1)
+    "Mass fraction of water in saturated moist air at the water inlet temperature";
+  Modelica.SIunits.SpecificEnthalpy hSatTWatIn=
+    Buildings.Media.Air.specificEnthalpy_pTX(
+      p=MediumA.p_default, T=TWatIn, X={X_wSatTWatIn,1-X_wSatTWatIn})
+    "Enthalpy of saturated moist air at the water inlet temperature";
+
+  Modelica.SIunits.AbsolutePressure pSatTWatOut=
+    Buildings.Utilities.Psychrometrics.Functions.saturationPressure(TWatOut)
+    "Saturation pressure of water at the water oulet temperature";
+  Modelica.SIunits.MassFraction X_wSatTWatOut=
+    Buildings.Utilities.Psychrometrics.Functions.X_pSatpphi(
+      pSat=pSatTWatOut, p=MediumA.p_default, phi=1)
+    "Mass fraction of water in saturated moist air at the water outlet temperature";
+  Modelica.SIunits.SpecificEnthalpy hSatTWatOut=
+    Buildings.Media.Air.specificEnthalpy_pTX(
+      p=MediumA.p_default, T=TWatOut, X={X_wSatTWatOut,1-X_wSatTWatOut})
+    "Enthalpy of saturated moist air at the water oulet temperature";
+
+  parameter Modelica.SIunits.SpecificEnthalpy LMED(fixed=false)
+    "Log mean enthalpy difference";
+
+  parameter Modelica.SIunits.MassFlowRate UASta(fixed=false, min=0)
+    "Overall heat transfer coefficient for enthalpy difference";
+  parameter Modelica.SIunits.HeatFlowRate QTot_flow=mWat_flow*cpw*(TWatOut-TWatIn)
+    "Total heat flow from air to water stream";
+
+  parameter Modelica.SIunits.ThermalConductance UAAir(min=0,start=10,fixed=false)
+  "Air side convective heat transfer coefficient, including fin resistance";
+  parameter Modelica.SIunits.ThermalConductance UAWat(min=0,start=20,fixed=false)
+  "Water side convective heat transfer coefficient";
+
+  parameter Boolean IsFulDry(fixed=false);
+  parameter Boolean IsFulWet(fixed=false);
+
+initial equation
+
+  if not use_UA_nominal then
+    assert(TAirOut<TAirIn and TWatOut>TWatIn and TWatIn<TAirIn,
+      "The rated condition is not for a cooling coil. " +
+      "For a heating coil, use Buildings.Fluid.HeatExchangers.DryCoilEffectivenessNTU.");
+
+    hAirIn=MediumA.specificEnthalpy_pTX(
+      p=MediumA.p_default, T=TAirIn, X={X_wAirIn, 1-X_wAirIn});
+    hAirOut=MediumA.specificEnthalpy_pTX(
+      p=MediumA.p_default, T=TAirOut, X={X_wAirOut, 1-X_wAirOut});
+
+    QTot_flow = mAir_flow*(hAirIn-hAirOut);
+
+    IsFulDry=(X_wSatTWatIn>=X_wAirIn);
+    IsFulWet=(X_wSatTWatOut<=X_wAirIn);
+
+    assert(IsFulDry or IsFulWet,
+      "In " + getInstanceName() + ":Cooling coil data under partially dry condition is not allowed at this moment. " +
+      "Specify either fully-dry or fully-wet nominal condition");
```

Comment:
This is not satisfactory for two reasons:

1. If this assertion triggers, in Dymola, the user gets a long list of all equations in this `if-then` construct, so it is hard to see what went wrong. This could be avoided by moving the calculation for `IsFulDry` and `IsFulWet` outside of the `if-then`, and then add a test for `use_UA_nominal` to the assertion conditions.
2. A partial wet coil at design conditions is probably not uncommon. In this case, the model only works if the user tricks the model by calculating `UA_nominal` outside of the model and assigning it to the parameter. But then, the same error would be done as if we were to compute `UA_nominal` inside the mode assuming a fully dry coil. Hence, maybe better would be to make a warning level for this assertion, and then proceed as if `IsFulDry` is `true`.

Reply:
In addition, when using `0.4*X_w_a2_nominal_wet/(1 - X_w_a2_nominal_wet)` for `w_a2_nominal` in `Buildings.Fluid.HeatExchangers.Validation.WetCoilEffectivenessNTU` to simulate partially wet nominal conditions, we get
```
Error: The following error was detected at time: 0

Model error - Modelica.Math.log (dT2/dT1) = Modelica.Math.log (-0.62475)

The stack of functions is:
Buildings.Fluid.HeatExchangers.BaseClasses.lmtd
Buildings.Fluid.HeatExchangers.BaseClasses.lmtd(44474.462595070174, 11526.990087133729, 19590.635318665023, 31567.470268785713)

FixInitials:Init: Integrator failed to start model.
```

Reply:
@mwetter I addressed the 2 items in fa199e3.
@kim1077 The model error reported in https://github.com/lbl-srg/modelica-buildings/pull/2364#discussion_r594670005 is still there: could you look into that?

Also, this assert statement seems not suitable either:
https://github.com/lbl-srg/modelica-buildings/pull/2364/commits/fa199e3e28ca944659c60adbc142d6887ac2e7ff#diff-9345b42f71c8410c9075361f1e22275e5775ee507189fe0c416a73e875da341dR132-R134
The user may want to use the model for a change-over coil for instance, and parameterize it based on heating mode conditions. 
So I propose to 

- remove that statement,
- add a validation case with heating nominal conditions and a transition from heating to cooling mode.

Reply:
@AntoineGautier, I understand user may may to use this model both heating and cooling. However, as indicated by the model name (**Wet**-Coil xxx), I don't expect users will specify heating data. I would like to have a meeting between you and I sometime next week (Monday?) To finalized this coil model.

Reply:
I added a validation case in heating mode in 3c81ab8, see `Buildings.Fluid.HeatExchangers.Validation.WetCoilEffectivenessNTUHeating`.

- With `use_UA_nominal=false` the wet coil model computes a null heat flow rate.
- With `use_UA_nominal=true` the wet coil model throws an error:
```
Error: The following error was detected at time: 0

Model error - division by zero: (hexWetNTU_TX.hA.hA_2) / (2.646*hexWetNTU_TX.dryWetCalcs.cpAir) = (4240.9) / (0)

Error: Integrator failed to start model.
```
The reason I want to support heating conditions is twofold:

1. In system simulation, it could happen that the inlet air is actually colder (transiently) than the chilled water.
2. Change over systems are highly common, at least in recent buildings in Europe. So I anticipate this user story: the user configures the wet coil model with manufacturer cooling nominal conditions, which could likely correspond to partially-wet regime. That would trigger the warning about model uncertainty. The user could then fall back to use manufacturer heating nominal conditions for the coil, precisely to bypass this warning by providing fully-dry conditions. And now he gets an error...

Reply:
https://github.com/lbl-srg/modelica-buildings/pull/2364#discussion_r594670005
I will have a look at the model.
Buildings.Fluid.HeatExchangers.Validation.WetCoilEffectivenessNTU

Reply:
The error is from lack of regularizations in the LMTD model (Buildings.Fluid.HeatExchangers.BaseClasses.lmtd).
See https://github.com/lbl-srg/modelica-buildings/issues/1515#issuecomment-519333479
To eliminate this error, the LMTD model has to be revised.

Reply:
@kim1077 Thinking about it further, I think the best way would be to refactor `Buildings.Fluid.HeatExchangers.BaseClasses.WetCoilUARated` to compute UA based on epsilon and NTU ( as in `Buildings.Fluid.HeatExchangers.BaseClasses.PartialEffectivenessNTU`) instead of LMED. That would avoid the need for regularizing `Buildings.Fluid.HeatExchangers.BaseClasses.lmtd` (which is tedious) and also take into account the various heat exchanger configurations (not only counter flow).

Reply:
Antoine, I just pushed the code we discussed today. Check that.